presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java

```diff
@@ -87,49 +92,23 @@ private static SchemaTableName parseTableName(String tableName)
     {
         SchemaTableName schemaTableName = new SchemaTableName(table.getDatabaseName(), table.getTableName());
 
-        List<LocatedFileStatus> files = cache.getIfPresent(path);
+        List<HiveFileInfo> files = cache.getIfPresent(path);
         if (files != null) {
-            return new HiveFileIterator(path, p -> new HadoopFileInfoIterator(simpleRemoteIterator(files)), namenodeStats, nestedDirectoryPolicy, pathFilter);
-        }
-        RemoteIterator<LocatedFileStatus> iterator = null;
-        try {
-            iterator = fileSystem.listLocatedStatus(path);
-        }
-        catch (IOException e) {
-            throw new PrestoException(HIVE_FILE_NOT_FOUND, "Hive file location does not exist: " + path);
+            return files.iterator();
         }
 
-        if (!cachedTableNames.contains(schemaTableName)) {
-            HadoopFileInfoIterator fileIterator = new HadoopFileInfoIterator(iterator);
+        if (cachedTableNames.contains(schemaTableName)) {
+            RemoteIterator<LocatedFileStatus> iterator = null;
+            try {
+                iterator = fileSystem.listLocatedStatus(path);
```

Comment:
I was imagining if we could do the following:

```
Iterator<HiveFileInfo> iterator = delegate.list(...);
if (cachedTableNames.contains(schemaTableName)) {
  return cachingIterator(iterator, path); 
}
return iterator;
```

And then we also change `cachingRemoteIterator` to:
```
Iterator<HiveFileInfo> cachingIterator(Iterator<HiveFileInfo> iterator, Path path) {
  return new Iterator<HiveFileInfo>() {
    private final List<HiveFileInfo> files = new ArrayList<>();
    ...
  };
}
```

The reason I'm asking is that the current design does not really go through the delegate `DirectoryLister` when we need to leverage cache. Let me know what you think :) Besides this, all looks good!

Reply:
oh, nice. Yeah, it is cleaner